chain/chain/src/thread_pool.rs

```diff
@@ -0,0 +1,107 @@
+}
+
+type Job = Box<dyn FnOnce() + Send + 'static>;
+type IdleQueue = Arc<Mutex<VecDeque<oneshot::Sender<Option<Job>>>>>;
+
+pub(crate) struct ThreadPool<C: ThreadLimit> {
+    name: &'static str,
+    limit: C,
+    idle_timeout: Duration,
+    num_threads: AtomicUsize,
+    idle_queue: IdleQueue,
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum Error {
+    #[error("thread limit reached for pool {name}: {limit}")]
+    ThreadLimitReached { name: &'static str, limit: usize },
+    #[error("spawning thread failed: {0}")]
+    Spawn(#[from] std::io::Error),
+}
+
+impl<C: ThreadLimit> ThreadPool<C> {
+    pub(crate) fn new(name: &'static str, idle_timeout: Duration, limit: C) -> Self {
+        Self {
+            name,
+            limit,
+            idle_timeout,
+            num_threads: Default::default(),
+            idle_queue: Default::default(),
+        }
+    }
+
+    pub(crate) fn spawn(&self, f: impl FnOnce() + Send + 'static) -> Result<(), Error> {
+        self.spawn_boxed(Box::new(f))
+    }
+
+    pub(crate) fn spawn_boxed(&self, job: Job) -> Result<(), Error> {
+        // Try to use one of the existing idle threads
+        let mut job = Some(job);
+        let mut idle_guard = self.idle_queue.lock();
+        while let Some(sender) = idle_guard.pop_front() {
+            job = match sender.send(job) {
+                Ok(()) => return Ok(()),
+                Err(err) => {
+                    // Idle thread has been terminated
+                    self.num_threads.fetch_sub(1, Ordering::Relaxed);
+                    err.into_inner()
+                }
+            }
+        }
+        drop(idle_guard);
+        self.try_spawn_thread(job.unwrap())
+    }
+
+    fn try_spawn_thread(&self, job: Job) -> Result<(), Error> {
+        let name = self.name;
+        let limit = self.limit.max_threads();
+        if self.num_threads.load(Ordering::Relaxed) >= limit {
+            return Err(Error::ThreadLimitReached { name, limit });
+        }
+
+        let idle_timeout = self.idle_timeout;
+        let idle_queue = self.idle_queue.clone();
+        ThreadBuilder::default()
+            .name(name)
+            .policy(ThreadSchedulePolicy::Realtime(RealtimeThreadSchedulePolicy::RoundRobin))
+            // TODO: Set thread priority
+            .spawn_careless(move || run_worker(job, idle_timeout, idle_queue))?;
+        self.num_threads.fetch_add(1, Ordering::Relaxed);
+
+        Ok(())
+    }
+}
+
+fn run_worker(mut job: Job, idle_timeout: Duration, idle_queue: IdleQueue) {
+    loop {
+        job();
+        // Notify the pool that this thread is idle by pushing the sender into the idle queue
+        let (sender, receiver) = oneshot::channel();
+        idle_queue.lock().push_back(sender);
+
+        job = match receiver.recv_timeout(idle_timeout) {
+            Ok(Some(job)) => job,
+            _ => break,
+        }
+        // No need to notify the pool that this thread has terminated – dropping the receiver is enough
```

Comment:
Do we actually need the support for idle thread and termination on the first attempt? I'd rather start easy if there isn't any data backing this up.

Reply:
Well, it's already implemented, so maybe leave it like that?

Reply:
As a rule I'd rather stick to the simplest solution, added complexity tends to accumulate over time. It's not a big deal though, If you do have a reason for keeping it, even just because you like it, feel free to do so.

Reply:
nit: I think this comment should be moved one line below.